### ISSUE TITLE
Rule proposal: `prefer-top-level-await`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#981]
Prefer top-level await instead of async IIFE or a top-level function call to an async function.

https://github.com/tc39/proposal-top-level-await

Top-level await is supported in Node.js since Node.js v14.3.0. So we cannot yet add it to the recommended preset, but we can add a TODO comment.

The async IIFE pattern is quite popular in command-line apps: https://github.com/sindresorhus/is-online-cli/blob/fb03be268bdec08e264dad31c5e7a110579d02f6/cli.js#L26-L33

## Fail

```js
(async () => {
	await getUnicorns();
})();
```

```js
function main() {
	await getUnicorns();
}

main();
```

## Pass

```js
await getUnicorns();
```

**Comment by Cherry** [comment]
This one would be awesome. It's worth noting that this is only supported in ES Modules with `"type": "module"` (or `.mjs`) though, so care would need to be taken not to flag old CJS usage as an error.

**Comment by sindresorhus** [comment]
> It's worth noting that this is only supported in ES Modules with "type": "module" (or .mjs) though, so care would need to be taken not to flag old CJS usage as an error.

We would not automatically detect this. It would be up to the user to decide whether they want the rule or not depending on what `type` they use.

**Comment by silverwind** [comment]
This may be tricky to detect. My CJS boilerplate for async modules is below, I guess the rule should be able to detect it as an error:

````js
function exit(err) {
  if (err) console.error(err);
  process.exit(err ? 1 : 0);
}

async function main() {
  // await stuff goes here
}

main().then(exit).catch(exit);

**Comment by sindresorhus** [comment]
@silverwind I think we would still be able to report on it, just not auto-fix it.

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by fisker** [comment]
Stage 4 now https://github.com/tc39/proposals/commit/99fa42a0863b1915eccead5bd2939ed5f468caab

**Comment by fisker** [comment]
Summary,


1. Top level Async IIFE

```js
(async function () {})();
```
```js
(async () => {})();
```

2. Top level call
```js
fn() // `fn` is an async function
```

3. Top level promise

```js
anything.{then,catch,finally}()
```

include

```js
(async () => {}).catch()
```

Enough?

**Comment by mvorisek** [comment]
how to fix https://github.com/fomantic/Fomantic-UI/blob/8a8d841a3eec3b1a103bb69be0350f1d7715c775/scripts/nightly-version.js#L96?

error: `96:6  error  Prefer top-level await over using a promise chain  unicorn/prefer-top-level-await`

maybe a another "pass" example can be added to the docs - https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-top-level-await.md
